918 Added tags and notes to docket alerts

albertisfu opened the pull request:
Added tags and notes to docket alerts as required in #918 

- Added a new method `get_docket_notes_and_tags_by_user` to get the user notes and tags for a docket and add them to the docket alert email.
- For the HTML version we show the tags with links, so it was needed to add the `username` to the email context in order to to generate the tag URL.
- For the plain text version, we only show the tag name.

- Added documentation for Favorites on the same page of Tags, now `/help/tags-favorites/`
- Docket alert screenshot updated in `/help/alerts/` to show the latest additions, Your notes, and Your Tags.

Here are some screenshots:

Docket with Notes and Tags:
![Screen Shot 2022-12-29 at 21 27 04](https://user-images.githubusercontent.com/486004/210031474-bdf49cce-b783-4229-8b81-eebad78e2e0d.png)

Docket with Tags but no Notes
![Screen Shot 2022-12-29 at 21 26 43](https://user-images.githubusercontent.com/486004/210031500-e2c6d58b-3d24-404a-b050-993e66a7a2b7.png)

Docket with Notes but no Tags
![Screen Shot 2022-12-29 at 21 25 33](https://user-images.githubusercontent.com/486004/210031521-691ff7a4-7989-4ac2-9418-b0020544dcd4.png)

Docket without Tags and Notes
![Screen Shot 2022-12-29 at 21 26 00](https://user-images.githubusercontent.com/486004/210031532-7cb89b8c-c807-4593-89f6-bca46251055b.png)

The `Learn more` link goes to `/help/tags-favorites/`


- Fixed the subject of `webhook_still_disabled_email`, I noticed it lacked the word "webhook" before the event type.
- Fixed the `test_send_search_alert_webhooks_rates` that checks all the rates for sending search alerts. And monthly alerts cannot be run on the 29th, 30th, or 31st, so it was failing (since today is 29th) added `time_machine` to mock the date.

Let me know what you think.

mlissner commented on cl/alerts/templates/docket_alert_email.html:
```diff
@@ -60,6 +60,28 @@ <h3 class="alt bottom" style="font-size: 1.5em; font-weight: normal; line-height
+        </span>
+      </p>
+    {% else %}
+      <p style="font-weight: inherit; font-family: inherit; border: 0; vertical-align: baseline; font-style: inherit; margin: 1em 0 1em; padding: 0;">
+        <strong>Your Note and Tags: </strong><span>You have not added a notes or tags to this alert. <a href="https://www.courtlistener.com{% url 'tag_favorites_help' %}">Learn more</a></span>
+      </p>
```
I think we can simplify in this case:

```suggestion
      <p style="font-weight: inherit; font-family: inherit; border: 0; vertical-align: baseline; font-style: inherit; margin: 1em 0 1em; padding: 0;">
        Use notes and tags to organize and share the cases you follow. <a href="https://www.courtlistener.com{% url 'tag_favorites_help' %}">Learn more</a>
      </p>
```

albertisfu replied:
Changed.

mlissner commented on cl/alerts/templates/docket_alert_email.txt:
```diff
@@ -20,6 +20,14 @@ CourtListener Docket Alert
+
+Your Tags: {% if tags %}{% for tag in tags %}{{ tag.name }}{% if not forloop.last %}, {% endif %}{% endfor %}{% else %} No tags yet. Learn More: https://www.courtlistener.com{% url 'tag_favorites_help' %}{% endif %}
+{% else %}
+Your Note and Tags: You have not added a notes or tags to this alert. Learn More: https://www.courtlistener.com{% url 'tag_favorites_help' %}
```
```suggestion
Use notes and tags to organize and share the cases you follow. Learn More: https://www.courtlistener.com{% url 'tag_favorites_help' %}
```

albertisfu replied:
Changed.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -11,26 +11,40 @@
-    <p>Currently you are only able to tag dockets, but we plan to add documents and other types of records in the future.</p>
-    <p>We hope that if you find our tags useful, you will donate to <a href="https://free.law">Free Law Project</a>, the non-profit supporting CourtListener and RECAP, to support the ongoing development and maintenance of this and other features.
-    </p>
+    <h1>Help with Tags and Favorites on CourtListener</h1>
```
I think I liked how this was before, but we can tweak it to also refer to favorites so:

```suggestion
    <h1>Help with Tags and Favorites on CourtListener</h1>
    <p>The tagging and favorites features on CourtListener help you to tag, organize, and share cases that you are working with.
    </p>
    <p>We hope that if you find these tools useful, you will donate to <a href="https://free.law">Free Law Project</a>, the non-profit supporting CourtListener and RECAP, to support the ongoing development and maintenance of these and other features.
    </p>
```

albertisfu replied:
Perfect, this is changed.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
     <p>Please stay tuned!</p>
+
+    <h2 id="favorites">Favorites</h2>
+    <p>You can use the Favorites feature to save Dockets, RECAP Documents, Opinions, and Oral arguments within your Favorites list. You can add notes to each favorite item as well.</p>
```
```suggestion
    <p>You can use the Favorites feature to save Dockets, RECAP Documents, Opinions, and Oral arguments to the <a href=LINK HERE>Favorites tab</a> in your account. You can add notes to each favorite item as well. Notes on dockets will be shown in <a href="LINK HERE">docket alerts</a> that we send you, if you subscribe to a case.
    </p>
```

albertisfu replied:
Changed and added links.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
+    <h2 id="favorites">Favorites</h2>
+    <p>You can use the Favorites feature to save Dockets, RECAP Documents, Opinions, and Oral arguments within your Favorites list. You can add notes to each favorite item as well.</p>
+    <h3 id="making-favorites">Favorite an item</h3>
+    <p>To favorite an item on CourtListener, you must be logged in and go to the Docket, RECAP Document, Opinion, or Oral Argument that you want to add to your favorites list.</p>
+    <p>Then click on the star icon located to the left of the item title.</p>
```
```suggestion
    <p>To favorite an item on CourtListener, log in and click the star icon to the left of the title of a Docket, RECAP Document, Opinion, or Oral Argument
    </p>
```

albertisfu replied:
Changed.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
+    <p>Complete the box with following information:</p>
+    <ol>
+      <li>Name: Required field, by the default the item title, you can change it.</li>
+      <li>Notes: Optional, a note to the favorite, until 500 characters.</li>
```
```suggestion
      <li>Notes: Optional, a private note explaining why the item is important or relevant (500 character limit).</li>
```

albertisfu replied:
Changed.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
+    </ol>
+    <p>Click on the save button.</p>
+    <h3 id="editing-favorites">Viewing, Editing or Deleting a Favorite</h3>
+    <p>You can view your favorite items in the Favorites tab within your user profile panel.</p>
```
Add a link here, please.

albertisfu replied:
Link added.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
+         alt="screenshot of the favorites tab"
+         class="img-responsive img-rounded shadow center-block"
+         height="503"
+         width="1164">
```
This width seems really wide. This should be set to the max width of the container.

albertisfu replied:
Yeah, I have fixed the with and height accordingly to the max width of the container.

mlissner commented on cl/simple_pages/templates/help/tags_help.html:
```diff
@@ -69,9 +84,52 @@ <h2 id="editing-tags">Editing or Deleting a Tag</h2>
+         height="519"
+         width="1164">
+    </p>
+    <p>You can change the content of the favorite and Save or delete it by clicking the “Delete” button in the bottom left corner.</p>
```
```suggestion
    <p>You can change the content of the favorite and save or you can delete it by clicking the "Delete" button in the bottom left corner.</p>
```

albertisfu replied:
Changed.

mlissner reviewed (changes requested):
These emails look great. I think this will finally drive some adoption among our users of these features. 

A handful of comments, and then a couple things:

1. For alert-example.png, you changed the dimensions, to make it wider and taller. The height will need to be adjusted in the HTML where it's viewed, and the width should *probably* be the same as before (but it's worth checking the max-width of the container where it's in, and setting it to that). 

4. For all your screenshots, can you update the note to say something like this, instead of Lorem Ipsum:

    > This the case about black lung personal responsibility among minor miners. The plaintiffs, a small class of coal minors, believe that minors should not be responsible for the damage to their lungs. The defendants claim teenage vaping as a defense and that the parents of all minors signed waivers on their behalf.

Thanks.

albertisfu commented on cl/alerts/tasks.py:
```diff
@@ -133,6 +137,35 @@ def get_docket_alert_recipients(
+    ).select_related("tag")
+    for docket_tag in docket_tags:
+        tags.append(docket_tag.tag)
+    return notes, tags
```
Great, yeah I agree about using the simplified query, so I changed it, thanks.

albertisfu commented:
Seems there is a mypy, error, checking it...

albertisfu commented:
These changes are completed. Also updated screenshots were there is a Note, using the following text:

> This the case about black lung personal responsibility among minor miners. The plaintiffs, a small class of coal minors, believe that minors should not be responsible for the damage to their lungs. The defendants claim teenage vaping as a defense and that the parents of all minors signed waivers on their behalf.

And fixed screenshots size in the HTML.

Let me know if there is anything else here.

Thank you!

mlissner commented on cl/alerts/tasks.py:
```diff
@@ -133,6 +137,33 @@ def get_docket_alert_recipients(
+        notes = favorite.notes
+
+    tags = UserTag.objects.filter(user_id=user_pk, dockets__id=d_pk)
+    for tag in tags:
+        user_tags.append(tag)
```
Why do we need this?

albertisfu replied:
yeah, this is not necessary, now I just converted the `QuerySet` to a `list()` since it's simpler to compare in the test.
Thanks!

mlissner reviewed (commented):
Looks good. One comment for you, below. 

I also noticed a couple typos in the screenshots. If it's easy to re-do them (like, less than five minutes), this text should be better (sorry):

> This case is about laughing disease liability among minor miners. The plaintiffs, a small class of coal miners, believe that minors should not be responsible for the unstoppable laughing caused by gases in the mines. The defendants claim teenage vaping as a defense and that the parents of all minor miners signed a waiver on their behalf.

albertisfu commented:
Yeah, I've applied the requested change and replaced the text in the screenshots.
Thanks!

mlissner commented:
Thanks, this is great!
